[Merged by Bors] - Add maxperf build profile

michaelsproul opened the pull request:
## Proposed Changes

Add a new Cargo compilation profile called `maxperf` which enables more aggressive compiler optimisations at the expense of compilation time.

Some rough initial benchmarks show that this can provide up to a 25% reduction to run time for CPU bound tasks like block processing: https://docs.google.com/spreadsheets/d/15jHuZe7lLHhZq9Nw8kc6EL0Qh_N_YAYqkW2NQ_Afmtk/edit

The numbers in that spreadsheet compare the `consensus-context` branch from #3604 to the same branch compiled with the `maxperf` profile using:

```
PROFILE=maxperf make install-lcli
```

## Additional Info

The downsides of the maxperf profile are:

- It increases compile times substantially, which will particularly impact low-spec hardware. Compiling `lcli` is about 3x slower. Compiling Lighthouse is about 5x slower on my 5950X: 17m 38s rather than 3m 28s.

As a result I think we should not enable this everywhere by default.

- **Option 1**: enable by default for our released binaries. This gives the majority of users the fastest version of `lighthouse` possible, at the expense of slowing down our release CI. Source builds will continue to use the default `release` profile unless users opt-in to `maxperf`.
- **Option 2**: enable by default for source builds. This gives users building from source an edge, but makes them pay for it with compilation time. 

I think I would prefer Option 1. I'll try doing some benchmarking to see how long a maxperf build of Lighthouse would take on GitHub actions.

Credit to Nicholas Nethercote for documenting these options in the Rust Performance Book: https://nnethercote.github.io/perf-book/build-configuration.html.

michaelsproul committed:
Add maxperf build profile

michaelsproul committed:
Don't strip binary

michaelsproul committed:
Enable maxperf for cross builds

michaelsproul committed:
Fix binary paths in release workflow

michaelsproul commented:
Interestingly, the release builds on CI don't seem to take much longer than they did previously:

- `maxperf` build on my fork: [1h 3m](https://github.com/michaelsproul/lighthouse/actions/runs/3132939057/jobs/5086341181)
- `release` build for v3.1.2: [42m](https://github.com/sigp/lighthouse/actions/runs/3124905197/jobs/5068730325)

I think this is because Github actions is already pretty close to single-threaded, so it wasn't gaining much from compiling the early crates in parallel (unlike the 5950X which gains _a lot_ by parallelism).

This has solidified my commitment to maxperf-by-default for binaries, and the faster `release` build by default for source builds (i.e. Option 1).

michaelsproul committed:
maxperf docker images

paulhauner reviewed (approved):
Nice find! :rocket:

paulhauner commented:
Oops, accidental submission. I agree with option 1 for the time being :relaxed: Approval still stands.

michaelsproul commented:
Cool! I think I've got the release CI pretty much sorted, but would like to do a practice run before we do a release for real. Maybe just a "pre-release" that we release for beta testing, something like `v3.1.2-maxperf-beta`?

michaelsproul committed:
Refine defaults & plumbing

michaelsproul committed:
Fix docker

michaelsproul commented:
Updated the book and tweaked the defaults so that:

- Running a cross build via the makefile defaults to `release` unless `CROSS_PROFILE=maxperf` is set. I figure this keeps us nimble while testing out changes. It saves like 15 minutes per build.
- Updated the `cross` builds for Docker and Linux releases to set `CROSS_PROFILE=maxperf`.
- Enabled the `maxperf` profile for Windows and Mac release workflows.

Running this on my fork here:

- Docker: https://github.com/michaelsproul/lighthouse/actions/runs/3147257075
- Release: https://github.com/michaelsproul/lighthouse/actions/runs/3147257088

Will merge once a few of those builds succeed.

michaelsproul commented:
bors r+

michaelsproul commented:
bors r-

michaelsproul commented:
bors r+

michaelsproul commented:
bors r-

michaelsproul commented:
bors r+
